cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -154,59 +155,48 @@ public void updateAppSpec(String namespaceId, String appId, ApplicationSpecifica
     write(key, updated);
   }
 
-  private void updateWorkflowRecord(Id.Program program, String pid, Map<String, String> systemArgs,
-                                    ProgramRunStatus status, @Nullable Throwable failureCause) {
-
-    String workflowNodeId = systemArgs.get(ProgramOptionConstants.WORKFLOW_NODE_ID);
-    String workflowName = systemArgs.get(ProgramOptionConstants.WORKFLOW_NAME);
-    String workflowRun = systemArgs.get(ProgramOptionConstants.WORKFLOW_RUN_ID);
-
-    ProgramRunId workflowRunId = new ProgramRunId(program.getNamespaceId(), program.getApplicationId(),
-                                                  ProgramType.WORKFLOW, workflowName, workflowRun);
+  public Map<String, WorkflowNodeStateDetail> getWorkflowNodeStates(ProgramRunId workflowRunId) {
+    MDSKey key = getProgramKeyBuilder(TYPE_WORKFLOW_NODE_STATE, workflowRunId.getParent().toId())
+      .add(workflowRunId.getRun()).build();
 
-    // Get the run record of the Workflow which started this program
-    MDSKey key = getWorkflowRunRecordKey(workflowRunId.getParent().toId(), workflowRunId.getRun());
+    List<WorkflowNodeStateDetail> nodeStateList = list(key, WorkflowNodeStateDetail.class);
 
-    RunRecordMeta record = get(key, RunRecordMeta.class);
-    if (record == null) {
-      String msg = String.format("No meta found for associated Workflow %s run record %s, while recording run for the" +
-                                   " namespace %s app %s type %s program %s runid %s", workflowRunId.getProgram(),
-                                 workflowRunId.getRun(), program.getNamespaceId(), program.getApplicationId(),
-                                 program.getType().name(), program.getId(), pid);
-      LOG.error(msg);
-      throw new IllegalArgumentException(msg);
+    Map<String, WorkflowNodeStateDetail> nodeStates = new HashMap<>();
+    for (WorkflowNodeStateDetail nodeStateDetail : nodeStateList) {
+      nodeStates.put(nodeStateDetail.getNodeId(), nodeStateDetail);
```

Comment:
Returning `List<WorkflowNodeStateDetail` from here. However it would be convenient for the users of this method to return map, for example from REST handler, its better to return the map so that user of handler can do look up based on the node name.